src/utils/robust.hpp

```diff
@@ -60,6 +60,24 @@ KOKKOS_INLINE_FUNCTION auto ratio(const A &a, const B &b) {
   const B sgn = b >= 0 ? 1 : -1;
   return a / (b + sgn * SMALL<B>());
 }
+
+KOKKOS_FORCEINLINE_FUNCTION
+bool soft_equiv(const Real &val, const Real &ref,
+                const Real eps = 10. * std::numeric_limits<Real>::epsilon(),
+                const bool pass_on_small = true) {
+  if (std::fabs(val - ref) < eps * fabs(ref)) {
+    // Return true equivalence if value and reference differ by less than precision
+    return true;
+  } else {
+    if (std::fabs(ref) < std::numeric_limits<Real>::min() && pass_on_small) {
+      // Optionally return true if reference value is close to zero
+      return true;
+    } else {
+      return false;
+    }
+  }
+}
```

Comment:
Is this performant enough to be used on device?

Reply:
I assume this is mostly intended for testing, so performance probably not too important.

That said, is branching actually necessary here? Can this not be written as a compound boolean? The latter almost certainly is performant enough to be used on device, and compound expressions are lazily evaluated. So could do...

```suggestion
// Return true equivalence if value and reference differ by less than precision
// Optionally return true if reference value is close to zero
KOKKOS_FORCEINLINE_FUNCTION
bool SoftEquiv(const Real &val, const Real &ref,
                const Real eps = 10. * std::numeric_limits<Real>::epsilon(),
                const bool pass_on_small = true) {
  return ((std::fabs(val - ref) < eps * std::fabs(ref)) 
    || ((std::fabs(ref) <std::numeric_limits<Real>::min()) && pass_on_small));
}
```
Also changed name to parthenon style ;)

Reply:
Good point @Yurlungur, I adopted your changes. This is is actually a use case for this in a compute intensive loop so this is a worthwhile change. Thanks also for catching the name style issue.

Reply:
As discussed on MM to make this more readable, you could also give the bools names, something like:
```C++
// Return true equivalence if value and reference differ by less than precision
// Optionally return true if reference value is close to zero
KOKKOS_FORCEINLINE_FUNCTION
bool SoftEquiv(const Real &val, const Real &ref,
                const Real eps = 10. * std::numeric_limits<Real>::epsilon(),
                const bool pass_on_small = true) {
  const bool is_close = (std::fabs(val - ref) < eps * std::fabs(ref);
  const bool is_small = std::fabs(ref) <std::numeric_limits<Real>::min();
  return (is_close || (is_small && pass_on_small));
}
```
which is a bit more readable I think.

Also it just occurred to me: Should we check that both the value and the reference are small for `pass_on_small`? If so we could do:
```C++
// Return true equivalence if value and reference differ by less than precision
// Optionally return true if reference value is close to zero
KOKKOS_FORCEINLINE_FUNCTION
bool SoftEquiv(const Real &val, const Real &ref,
                const Real eps = 10. * std::numeric_limits<Real>::epsilon(),
                const bool pass_on_small = true) {
  constexpr Real tiny = std::numeric_limits<Real>::min();
  const bool is_close = (std::abs(val - ref) < eps * std::abs(ref);
  const bool ref_is_small = std::abs(ref) < tiny;
  const bool val_is_small = std::abs(val) < tiny;
  const bool is_small = ref_is_small && val_is_small;
  return (is_close || (is_small && pass_on_small));
}
```
also I *think* that in C++ `fabs` and `abs` are templated to have identical behavior for identical inputs? Not sure?

Reply:
You're right about `fabs` -- that's just my C99 instincts. Also agree with the `const bool is_` approach, nice suggestion.

I think the logic behind `pass_on_small` is that if the reference value is `0`, the comparison isn't well-defined, so the person using this function can decide either to let those cases all pass or all fail. That isn't quite the same behavior as checking that both `val` and `ref` are small, and I think the former is more useful.

Reply:
Yeah I'm fine either way regarding the behavior for `pass_on_small` my main logic was if the reference value is tiny, the correct answer is the test value should be too. Either way though.